modules/transaction-pause/src/lib.rs

```diff
@@ -0,0 +1,132 @@
+// This file is part of Acala.
+
+// Copyright (C) 2020-2021 Acala Foundation.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+#![allow(clippy::unused_unit)]
+
+use frame_support::dispatch::{CallMetadata, GetCallMetadata};
+use frame_support::{pallet_prelude::*, traits::Filter, transactional};
+use frame_system::pallet_prelude::*;
+use sp_runtime::DispatchResult;
+use sp_std::{prelude::*, vec::Vec};
+
+mod mock;
+mod tests;
+pub mod weights;
+
+pub use module::*;
+pub use weights::WeightInfo;
+
+#[frame_support::pallet]
+pub mod module {
+	use super::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// The origin which may set filter.
+		type UpdateOrigin: EnsureOrigin<Self::Origin>;
+
+		/// Weight information for the extrinsics in this module.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		// can not pause
+		CannotPause,
+		// invalid character encoding
+		InvalidCharacter,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(fn deposit_event)]
+	pub enum Event<T: Config> {
+		TransactionPaused(Vec<u8>, Vec<u8>),
+		TransactionUnpaused(Vec<u8>, Vec<u8>),
+	}
+
+	/// The paused transaction map
+	///
+	/// map (PalletNameBytes, FunctionNameBytes) => Option<()>
+	#[pallet::storage]
+	#[pallet::getter(fn paused_transactions)]
+	pub type PausedTransactions<T: Config> = StorageMap<_, Twox64Concat, (Vec<u8>, Vec<u8>), (), OptionQuery>;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(T::WeightInfo::pause_transaction())]
+		#[transactional]
+		pub fn pause_transaction(origin: OriginFor<T>, pallet_name: Vec<u8>, function_name: Vec<u8>) -> DispatchResult {
+			T::UpdateOrigin::ensure_origin(origin)?;
+
+			// because the pallet name of this module in runtime is undetermined here,
+			// not allowed to pause calls have same names with these calls to ensure safe
+			// event if maybe there are calls with the same name in other pallet.
+			let function_name_string =
```

Comment:
it should be possible to get pallet name using `PalletInfoAccess`

https://github.com/paritytech/substrate/blob/71e1c59e7e901882487ee516cf2faa02aed74da5/frame/support/test/tests/pallet.rs#L1265